chacha20_poly1305/src/chacha20.rs

```diff
@@ -0,0 +1,419 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! The ChaCha20 stream cipher from RFC8439.
+
+/// The first four words (32-bit) of the ChaCha stream cipher state are constants.
+const WORD_1: u32 = 0x61707865;
+const WORD_2: u32 = 0x3320646e;
+const WORD_3: u32 = 0x79622d32;
+const WORD_4: u32 = 0x6b206574;
+
+/// The quarter round indexes of a ChaCha block function.
+const CHACHA_ROUND_INDICIES: [(usize, usize, usize, usize); 8] = [
+    // The first 4 rounds are columns of the 4x4 matrix.
+    (0, 4, 8, 12),
+    (1, 5, 9, 13),
+    (2, 6, 10, 14),
+    (3, 7, 11, 15),
+    // The next 4 rounds are diagonals of the 4x4 matrix.
+    (0, 5, 10, 15),
+    (1, 6, 11, 12),
+    (2, 7, 8, 13),
+    (3, 4, 9, 14),
+];
+
+/// The cipher's block size is 64 bytes.
+const CHACHA_BLOCKSIZE: usize = 64;
+
+/// A 256-bit secret session key shared by the parties communicating.
+#[derive(Clone, Copy, Debug)]
+pub struct SessionKey([u8; 32]);
+
+impl SessionKey {
+    /// Create a new session key.
+    pub const fn new(key: [u8; 32]) -> Self {
+        SessionKey(key)
+    }
+}
+
+/// A 96-bit initialization vector (IV), or nonce.
+#[derive(Clone, Copy, Debug)]
+pub struct Nonce([u8; 12]);
+
+impl Nonce {
+    /// Create a new nonce.
+    pub const fn new(nonce: [u8; 12]) -> Self {
+        Nonce(nonce)
+    }
+}
```

Comment:
How are these chosen? Random or counting? I think methods to generate a random or add 1 depending on which method is used would be nice. But shouldn't block the PR.

Reply:
Typically by counting. Probably it would be more useful to have constructors from integer types than to have methods that modify an existing nonce.

Reply:
Having weird length of 12B I think a counting method would be more appropriate since `from_u64` wastes some bits and `from_u128` is fallible. But ACK *also* having methods to pass `u64` and `u128`.

Reply:
It's 96 bits wide but typical usage is to just use a 64-bit counter and leave the high bits at 0 (or *sometimes* set them to some protocol-specific fixed value).

Reply:
I guess people are just lazy to implement those, right? :)

I mean I don't want to block the PR for this but I still think it'd be nice regardless.

Reply:
Actually, I realize I am confusing two things here. The `Nonce` is 96 bits and is input into the state alongside a 32-bit counter (and together the nonce and counter form a single 128-bit object which the algorithm treats as a single thing). So the intended use here seems to be for the nonce to be fixed and the `count` to change.

From the RFC:

> Note also that the original ChaCha had a 64-bit nonce and 64-bit
> block count.  We have modified this here to be more consistent with
> recommendations in Section 3.2 of [RFC5116].  This limits the use of
> a single (key,nonce) combination to 2^32 blocks, or 256 GB, but that
> is enough for most uses.  In cases where a single key is used by
> multiple senders, it is important to make sure that they don't use
> the same nonces.  This can be assured by partitioning the nonce space
> so that the first 32 bits are unique per sender, while the other 64
> bits come from a counter.

So my "64-bit counter, the rest zeros" is apparently outdated, and this paragraph is what the API should support.

Reply:
Good points. I'm not entirely sure how the API should look like but we ought to support it somehow.

Reply:
This interface (or lack there of) is totally influenced by the [wrapper used in BIP324 to derive nonces](https://github.com/bitcoin/bips/blob/master/bip-0324.mediawiki#user-content-Rekeying_wrappers_FSChaCha20Poly1305_and_FSChaCha20), which to me feels kinda special? So not this interface should adopt any thing to help it out. But I am not sure what would help other callers.

Reply:
If this is a general-purpose crate we should expose raw access to the nonces and implement the wrapper in the BIP crate. If there are no other implementors so far, it's OK to have just the raw API.

Reply:
That all makes sense to me. Is the API ok in its current spot, or do you want me to push it in one direction or another for now?

Reply:
I think it's fine.